kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/security/ElytronUserGroupAdapter.java

```diff
@@ -16,35 +16,90 @@
 
 package org.kie.server.services.jbpm.security;
 
-import java.util.Collections;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
 
 import org.jbpm.services.task.identity.adapter.UserGroupAdapter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wildfly.security.auth.server.RealmIdentity;
+import org.wildfly.security.auth.server.RealmUnavailableException;
 import org.wildfly.security.auth.server.SecurityDomain;
 import org.wildfly.security.auth.server.SecurityIdentity;
+import org.wildfly.security.authz.AuthorizationFailureException;
 
 public class ElytronUserGroupAdapter implements UserGroupAdapter {
 
+    private static final Logger logger = LoggerFactory.getLogger(ElytronUserGroupAdapter.class);
+
     @Override
     public List<String> getGroupsForUser(String userId) {
 
-        Optional<SecurityIdentity> identityOptional = getCurrentSecurityIdentity();
+        String userName = getUserName();
+        logger.debug("Identifier Elytron as {}", userId);
+        if (userName == null) {
+            return new ArrayList<>();
+        }
 
-        if (identityOptional.isPresent()) {
-            SecurityIdentity identity = identityOptional.get();
-            if (identity.getPrincipal().getName().equals(userId)) {
-                return StreamSupport.stream(identity.getRoles().spliterator(), false)
-                        .collect(Collectors.toList());
+        if (userId.equals(userName)) {
+            logger.debug("User identified as {} but auth as {}", userId, userName);
+            return toPrincipalRoles(userId);
+        } else {
+            try {
+                if (runAsPrincipalExists(userId)) {
+                    logger.debug("Executing run as {}", userId);
+                   return toRunAsPrincipalRoles(userId, true);
+                } else {
+                   return new ArrayList<>();
+                }
+            } catch (AuthorizationFailureException ex) {
+                return toRunAsPrincipalRoles(userId, false);
+            } catch (RealmUnavailableException | SecurityException e) {
+                return new ArrayList<>();
             }
         }
+    }
 
-        return Collections.emptyList();
+    public List<String> toPrincipalRoles(String userId) {
+        return toRoles(getCurrentSecurityIdentity().get());
+    }
+
+    public List<String> toRunAsPrincipalRoles(String userId, boolean authenticate) {
+        return toRoles(getCurrentSecurityIdentity().get().createRunAsIdentity(userId, authenticate));
```

Comment:
Same here

Reply:
this is already checked when getting the userNAme so not really needed here.